### PR TITLE
unbound: optionally support DNS-over-HTTPS

### PR DESCRIPTION
**Pull request by nagy**
###### Motivation for this change
unbound can be used as a DNS-over-HTTPS (DoH) server.

This is a blog post introducing the feature:

https://www.nlnetlabs.nl/news/2020/Oct/08/unbound-1.12.0-released/

###### Notify Maintainers
cc @ehmry @fpletz @globin

###### Things done
- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/networking/unbound/default.nix [hunk 1]
```diff
@@ -18,6 +18,9 @@
   #
 , withSystemd ? false
 , systemd ? null
+  # optionally support for DNS-over-HTTPS as a server
+, withDoH ? false
+, libnghttp2 ? null
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/unbound/default.nix:23]:
```suggestion
, libnghttp2
```

File: pkgs/tools/networking/unbound/default.nix [hunk 2]
```diff
@@ -31,7 +34,9 @@ stdenv.mkDerivation rec {
 
   outputs = [ "out" "lib" "man" ]; # "dev" would only split ~20 kB
 
-  buildInputs = [ openssl nettle expat libevent ] ++ lib.optionals withSystemd [ pkg-config systemd ];
+  buildInputs = [ openssl nettle expat libevent ]
+                    ++ lib.optionals withSystemd [ pkg-config systemd ]
+                    ++ lib.optionals withDoH [ libnghttp2 ];
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/unbound/default.nix:39]:
```suggestion
  buildInputs = [ openssl nettle expat libevent ]
    ++ lib.optionals withSystemd [ pkg-config systemd ]
    ++ lib.optionals withDoH [ libnghttp2 ];
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by SuperSandro2000** [comment]
Build on linux amd64.

File: pkgs/tools/networking/unbound/default.nix (hunk 2 above)
Comment (nagy) [reply pkgs/tools/networking/unbound/default.nix]: updated

File: pkgs/tools/networking/unbound/default.nix (hunk 1 above)
Comment (SuperSandro2000) [inline pkgs/tools/networking/unbound/default.nix:23]:
```suggestion
, libnghttp2
```

**Review by SuperSandro2000 — APPROVED** [review]
other than that LGTM

**Comment by mweinelt** [comment]
This looks good. 

Are you possibly interested in adding a small test case to the existing `nixos/tests/unbound.nix`? This would help us validate the functionality of this feature going forward.

**Review by mweinelt — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Comment by nagy** [comment]
> Are you possibly interested in adding a small test case to the existing `nixos/tests/unbound.nix`? This would help us validate the functionality of this feature going forward.

It would be my first nixos test but I can give it a try.

**Comment by nagy** [comment]
The introduced `withDoH` argument default to `false`, meaning that the feature itself would not be built without overriding the package.

The test in `nixos/test/unbound.nix` tests the default package ( which makes sense to me). Do you want to switch `withDoH` to `true by default or would you rather override the package in the testsuite ?

**Comment by andir** [comment]
> The introduced `withDoH` argument default to `false`, meaning that the feature itself would not be built without overriding the package.
> 
> The test in `nixos/test/unbound.nix` tests the default package ( which makes sense to me). Do you want to switch `withDoH` to `true by default or would you rather override the package in the testsuite ?

The NixOS module is using the `unbound-with-systemd` package by default. We could rename that package to `unbound-full` (or whatever) and add DoH support to it. The reason we wouldn't do this to the default `unbound` package is that is a common dependency for a ton of packages (including everything that somewhere in the tree depends on systemd or udev).

**Comment by nagy** [comment]
I have drafted out something which seems to pass. Do you have any suggested changes ?

**Pull request merged**